### PR TITLE
[WIP] Paramaterizing operations to make a more composable system

### PR DESCRIPTION
**Pull request by litch**
These are some first steps toward parameterizing the `optimize.py` routine - there is a lot of duplication across the `optimize`, `train` and `test` files, and I'm hoping that by using some class inheritance, we can generally make the differences between the operations more understandable, and the code in general more easily extended.

### CODE REVIEW

File: optimize.py [hunk 1]
```diff
@@ -20,106 +21,108 @@
...
-
-    return -1 * last_reward
-
-
-def optimize():
-    study_name = 'ppo2_' + reward_strategy
+class Optimize:
+  def __init__(self):
+    self.reward_strategy = 'sortino'
+    self.input_data_file = os.path.join('data', 'coinbase_daily.csv')
```
Comment (litch) [inline optimize.py]: I'm using the daily file for testing so that Ic an have a reasonable rate of feedback for myself.

**Comment by rhamnett** [comment]
This is looking good, various fixes which prevent current master working out of the box.

**Comment by litch** [comment]
Currently the log output looks like this:
```
☺ ~/code/Bitcoin-Trader-RL[master*]% python optimize.py
  2019-06-10 06:21:39,226 - __main__ - __init__ - DEBUG - Initialized Optimizer
  2019-06-10 06:21:39,226 - __main__ - optimize - INFO - Running built-in data preparation                                                                                2019-06-10 06:21:39,986 - __main__ - log_parameters - DEBUG - Reward Strategy: sortino
  2019-06-10 06:21:39,986 - __main__ - log_parameters - DEBUG - Input Data File: data/coinbase_daily.csv
  2019-06-10 06:21:39,986 - __main__ - log_parameters - DEBUG - Params DB File: sqlite:///params.db
  2019-06-10 06:21:39,986 - __main__ - log_parameters - DEBUG - Parallel jobs: 1                                                                                          2019-06-10 06:21:39,987 - __main__ - log_parameters - DEBUG - Trials: 1
  2019-06-10 06:21:39,987 - __main__ - log_parameters - DEBUG - Test episodes (per trial): 1
  2019-06-10 06:21:39,987 - __main__ - log_parameters - DEBUG - Evaluations (per trial): 1
  2019-06-10 06:21:39,987 - __main__ - log_parameters - DEBUG - Train DF Length: 1034                                                                                     2019-06-10 06:21:39,987 - __main__ - log_parameters - DEBUG - Test DF Length: 259
  2019-06-10 06:21:39,987 - __main__ - log_parameters - DEBUG - Features: index, Date, Open, High, Low, Close, Volume BTC, Volume USD, RSI, MFI, TSI, UO, AO, MACD_diff,
 Vortex_pos, Vortex_neg, Vortex_diff, Trix, Mass_index, CCI, DPO, KST, KST_sig, KST_diff, Aroon_up, Aroon_down, Aroon_ind, BBH, BBL, BBM, BBHI, BBLI, KCHI, KCLI, DCHI,
DCLI, ADI, OBV, CMF, FI, EM, VPT, NVI, DR, DLR                                                                                                                          

... A bunch of Tensorflow warnings from the stable_baselines library...

  2019-06-10 06:27:36,873 - __main__ - optimize - INFO - Number of finished trials: 46
  2019-06-10 06:27:36,873 - __main__ - optimize - INFO - Best trial:
  2019-06-10 06:27:36,916 - __main__ - optimize - INFO - Value: -6338.4296875
  2019-06-10 06:27:36,916 - __main__ - optimize - INFO - Params:
  2019-06-10 06:27:36,916 - __main__ - optimize - INFO -     cliprange: 0.303861773084457
  2019-06-10 06:27:36,916 - __main__ - optimize - INFO -     confidence_interval: 0.9817939576549914
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     ent_coef: 0.06726894670753417
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     forecast_len: 24.805106879968463
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     gamma: 0.9114240945572387
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     lam: 0.9991480714490228
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     learning_rate: 0.14830353137574787
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     n_steps: 650.5846214860246
  2019-06-10 06:27:36,917 - __main__ - optimize - INFO -     noptepochs: 1.6915777278434119
```

**Comment by litch** [comment]
@notadamking - Any feedback on the way the optimize file looks before I port over this "style" to the train and test files?

Also, I'm thinking that I will change the `optimize.py` file to use tho same defaults you had before (so it takes ages, runs a thousand trials, whatever), and then have the `bootstrap.py` file load up each of the scripts (optimize, train, test) and run them in "fast mode" for an intro.

This also gives us some "handles" for running multiple instruments (with the same system, or having a different system-per-instrument) and rewards without changing the "guts" of the code too much.

**Comment by notadamking** [comment]
@litch Sorry for the slow response, I've been held up with work all weekend. Looks great so far!

At a glance, The first thing I notice is that by wrapping this all in a class architecture, it now makes sense to combine all of the optimize/train/test functionality into a single class, and allow the user to pass parameters into these functions, rather than having to re-define the variables in `optimize.py`, `train.py`, and `test.py`. 

At some point we should also further abstract the `prepare_data` method into a `DataLoader` class that can be used to import and prepare data sets from multiple sources.

Great work, I look forward to merging this PR!

**Comment by litch** [comment]
Thanks!  I agree about the prepare_data -> DataLoader, and was really impressed in general with some of the work of @marvin-hansen around in the area.  Ok, I'll see what I can generalize a touch more, then we can work on making improvements like that.

**Comment by notadamking** [comment]
Great! Thanks for all your help with this repo. I look forward to seeing what you come up with.

On a side note, this might be a good time to implement the suggestions in #6, allowing using to pass in arguments such as the agent, policy, etc.

**Comment by litch** [comment]
As a side note, why, in the `train.py` file, do you set the curr_idx to -1, and then make an assignment on it outside of the loop where you are iterating that variable?  Was that mostly a testing/development process thing?

https://github.com/notadamking/Bitcoin-Trader-RL/blob/master/train.py#L53

**Comment by notadamking** [comment]
Yes, I just threw something together for testing and I'm sure there is a much better way to do that.

**Comment by litch** [comment]
Also, just for my understanding, it looks like in `train` you train the model 10x, and then pickle the model to `./agents/ppo2_sortino_<index>_.pkl`.  Then, in `test` you load up the version dictated by the `curr_idx` variable at the top of the file.  That is so that you could, say, manually run the `test` method for any of the versions of the trained model, and which instance of the model you evaluated was dictated by that `curr_idx` variable at the top.  

Does my understanding seem right there?  So if you were curious about how "model instance number 7" performs, you just change the `curr_idx` at the top of `test.py` and run `test.py`?

**Comment by rhamnett** [comment]
> Also, just for my understanding, it looks like in `train` you train the model 10x, and then pickle the model to `./agents/ppo2_sortino_<index>_.pkl`. Then, in `test` you load up the version dictated by the `curr_idx` variable at the top of the file. That is so that you could, say, manually run the `test` method for any of the versions of the trained model, and which instance of the model you evaluated was dictated by that `curr_idx` variable at the top.
> 
> Does my understanding seem right there? So if you were curious about how "model instance number 7" performs, you just change the `curr_idx` at the top of `test.py` and run `test.py`?

@litch that's what I have been doing yeah.

**Comment by notadamking** [comment]
Exactly @litch. You've got it.

**Comment by litch** [comment]
Ok.  made a bit of progress (I'm on the road right now so can only work in small batches). 

Todo:
1) Rename the `optimize.py` system/setup/framework/whatever
2) Make a new `optimize.py` file that will load the framework and run the optimize method (parameterize from command line?)
3) Make the `train.py` and `test.py` load and run the framework (params from command line?)

File: optimize.py [hunk 2]
```diff
@@ -165,13 +164,91 @@ def optimize(self):
...
+          action, _states = model.predict(obs)
+          obs, reward, done, info = test_env.step(action)
+          reward_sum += reward
+
+      self.logger.info('[', idx, '] Total reward: ', reward_sum, ' (' + self.reward_strategy + ')')
+      model.save(os.path.join('.', 'agents', 'ppo2_' + self.reward_strategy + '_' + str(idx) + '.pkl'))
+
+    self.logger.info("Trained {} model instances".format(models_to_train))
+
+  def test(self, model_instance: 0):
```
Comment (notadamking) [inline optimize.py]: This should be `model_instance: int = 0` or simply `model_instance = 0`.

**Comment by litch** [comment]
Also, it looks like we are using the same data set as a "validation set" and a "test set", which should  tend toward overfitting.  Have you done trials of this methodology where the first time the system ever sees the "test data" is at the "final test" phase?

**Comment by notadamking** [comment]
The way I've coded it may be a bit confusing, but we first split the data into 80/20, the last 20% being the test set. Then we take the first 80% and split it again 80/20, this time the last 20% is the validation set, and the first 80% is the training set. Does that make sense?

**Comment by litch** [comment]
Oh!  Ok, that does make sense.  I think that subtlety has not made it into my changes.  I'll make sure it does.

**Comment by notadamking** [comment]
Just so you know I will be out of the country for a week starting tomorrow, so I won't be around to merge this, but I will get it merged as soon as I can when I get back.

**Comment by litch** [comment]
Yeah I'm actually traveling in Southeast Asia right now for work so things are a bit disrupted on my end.

**Comment by notadamking** [comment]
Hey @litch, have you made any more progress on this since your last commit?

If not, no worries, either way I am going to pick up where you left off today and try to get it integrated into the library.

Feel free to join the discord: https://discord.gg/ZZ7BGWh
